AVL GTFS Adjustment & Scheduler

wsp-mbuckley opened the pull request:
Added functionality for calculating delays from AVL data and adjusting the scheduled GTFS data to produce an "adjusted" schedule.

Also, added the scheduler functionality for regularly updating the data and uploading to the database.

Includes all changes from #2

wsp-mbuckley committed:
Initial AVL modules

Added existing AVL modules created during process testing.

wsp-mbuckley committed:
Added pyproject.toml

Using some parameters from caf.toolkit's file.

wsp-mbuckley committed:
Added GTFS-rt parsing dataclasses

Added classes for parsing and storing GTFS-rt data.
Added config for downloader.

wsp-mbuckley committed:
AVL bounding box parameter

Added bounding box parameter to AVL downloads.
Added error checking / logging to request.get.
Added docstrings.

wsp-mbuckley committed:
GTFS-rt downloader loop

Added continuous downloading for a given period of time.

wsp-mbuckley committed:
GTFS dataclasses added get_fields methods

wsp-mbuckley committed:
Added _Database base class

Added base class for accessing database ready for setting up GTFS class.

wsp-mbuckley committed:
GTFS-rt database class

Added functionality for storing GTFS-rt feed data into SQLite database.

wsp-mbuckley committed:
GTFS-rt fixed parsing trip start date / time

Parsing start date / time as isoformat.

wsp-mbuckley committed:
GTFS-rt database summary script

Added script to produce summaries from GTFS-rt database.

wsp-mbuckley committed:
GTFS database summary timestamp comparison

Added config file.
Added histogram of timestamp differences.

wsp-mbuckley committed:
Fixed GTFS summary table decimals

Added 1d.p. to GTFS summary table percentages < 1%.

wsp-mbuckley committed:
GTFS AVL fixed trip ID empty string

Updated GTFS-rt TripDescriptor parsing to return empty trip ID strings
as None so they are null in the database.

wsp-mbuckley committed:
GTFS AVL easting and northing

Added easting and northing to positions dataclass.
Added pyproj requirement, although process will run without it but
easting and northing will be NaN.

wsp-mbuckley committed:
Moved scripts folder into src

wsp-mbuckley committed:
Commented python version in requirements

Pip / pipenv don't understand having python in requirements.txt.

wsp-mbuckley committed:
AVL stop estimates SQL tests

Testing SQL queries for estimating stop arrival time in AVL data.

wsp-mbuckley committed:
AVL conversion to easting / northing accuracy

Set minimum required accuracy of conversion to 5 metres, errors will be
logged for anything less accurate or any other tranformation errors.

wsp-mbuckley committed:
AVL - SQL distances calculation

wsp-mbuckley committed:
GTFS-rt parser wheelchair accessible

Added wheelchair_accessible to VehicleDescriptor.
This isn't currently implemented in gtfs-realtime-bindings v1.0.0 but is
defined in the GTFS-rt specification.

wsp-mbuckley committed:
Fixed AVL downloader finish message

Removed final waiting after end time and added new finish message.

wsp-mbuckley committed:
GTFS-rt database summary insert timings

Extract database row insert times from AVL log file and plot.

wsp-mbuckley committed:
AVL - Added position delay column

Added timestamp delay data to positions dataclass which is included in the database table.
Added SQL script to update the same column in an existing database.

wsp-mbuckley committed:
AVL summary positions count plot

Plot number of positions in response over time.
Added SQL script for summarising data by response, which will be included in the summary script.

wsp-mbuckley committed:
GTFS-rt summary of positions per response

Added functions to extract count of positions per response from database and create graphs.

wsp-mbuckley committed:
Merge remote-tracking branch 'origin/avl-init' into gtfs_adjustment

wsp-mbuckley committed:
AVL vehicle position hash

Added vehicle `identifier_hash` column with unique constraint so duplicate vehicle positions aren't included.

wsp-mbuckley committed:
AVL download remaining wait time

Calculate remaining time needed to wait to account for time taken to handle previous request and get closer to the actual `wait_time` between requests.

wsp-mbuckley committed:
AVL downloader hash set size limit

Clear hash storing set when it gets larger than 10MB so we don't run into any memory issues.

wsp-mbuckley committed:
AVL identifier hash column size

wsp-mbuckley committed:
AVL GTFS summary new log file format

Updated extracting insert timings from the log file to the new version of log messages, which include insert time taken.
Fixed vehicle position delay calculation.

wsp-mbuckley committed:
AVL downloader removing unique database constraint

Removing unique constraint on database `identifier_hash` column.
Hashes are still checked with in memory list of previous hashes and inserts won't be attempted for any already found, this won't prevent duplicates in the database but will reduce the number of rows inserted that have been seen in the current run of the tool.

wsp-mbuckley committed:
AVL Fixed vehicle position identifier hash

Removed position_delay_seconds from hash as this is calculated based on response timestamp.

wsp-mbuckley committed:
AVL delete duplicate positions

Added method for deleting duplicate positions from AVL database once download is complete.

wsp-mbuckley committed:
AVL vehicle speeds table

Added table of vehicle speeds and queries to calculate and fill it.

wsp-mbuckley committed:
AVL GTFS stop times extract

Added function for loading GTFS stop times and locations.
Updated AVL speed table to only include a single value for each trip stop per day.

wsp-mbuckley committed:
AVL adjust config and stop times

Added config for AVL adjust process.
Inserted stop times data into AVL database.
Renamed delay module to adjust.

wsp-mbuckley committed:
AVL database SQL queries

Added queries for processing the AVL data in the SQlite database, copies of these queries will be used within the Python code.

wsp-mbuckley committed:
GTFS adjust stop distances

Updated GTFS stop time schedule in database to calculate arrival and departure times in seconds and distance between current and previous stops.

wsp-mbuckley committed:
AVL delay estimates

Added delay estimates functionality.
Added filtering incorrect GPS positions.

wsp-mbuckley committed:
AVL data area filtering script

Added script for outputting AVL data for a specific area for testing / checking.

wsp-mbuckley committed:
Added scheduler module outline

wsp-mbuckley committed:
Merge remote-tracking branch 'origin/main' into gtfs_adjustment

wsp-mbuckley committed:
Timetable download function

Moved GTFS timetable download functionality into separate function for use in scheduler

wsp-mbuckley committed:
Downloading GTFS and uploading to database.

Added module for connecting to the postgresql database.
Updated scheduler to download and insert timetable from BODS.

wsp-mbuckley committed:
Added scheduler command to CLI

wsp-mbuckley committed:
Added bodse package __main__

wsp-mbuckley committed:
Fixed database timetable path

wsp-mbuckley committed:
Fixed CLI command choices

wsp-mbuckley committed:
Added Zoning system table

wsp-mbuckley committed:
Fixed database zone system reference

wsp-mbuckley committed:
Fixed sqlalchemy logging

Filtered sqlachemy log messages from console output.
Fixed inserting metadata.
Added on delete cascade to foreign key constraints.

wsp-mbuckley committed:
Fixed GTFS adjustment process

Finished off GTFS adjustment process and got it ready for integration into scheduler.

wsp-mbuckley committed:
Added GTFS adjustment to scheduler

wsp-mbuckley committed:
Added MS Teams functionality

Added module to post messages to MS Teams.
Added docstrings and updated requirements.

wsp-mbuckley committed:
Changed wait time to 12hrs

Added version to pymsteams.
Disabled sqlalchemy logging as it created too many messages.
Add scheduled run timings.
Added error handling in AVL downloads to continue to log any errors and continue.

wsp-mbuckley committed:
Scheduler calculated longer wait time

wsp-mbuckley committed:
Fixed parsing weekday for scheduler config

wsp-mbuckley committed:
Fixed scheduler config parsing day

wsp-mbuckley committed:
Fixed zoning system schema

wsp-mbuckley committed:
Fixed run_metadata schema

wsp-mbuckley committed:
Fixed bus timetables schema

wsp-mbuckley committed:
Fixed AVL download error handling

Fixed catching AVL download errors.
Added catching / logging teams post errors.

wsp-mbuckley committed:
Fixed pymsteams missing module warning

wsp-mbuckley committed:
Added log file path to teams error posts

wsp-mbuckley committed:
Added startup teams post

wsp-mbuckley committed:
Setup package install info

wsp-mbuckley committed:
Fixed stop times inserting

Added check to see if table exists before counting rows.
Added method for checking table exists to `_Database` base class.
Fixed error in `calculate_stop_times_delays` if output folder isn't given.

wsp-mbuckley committed:
Allowed multiple weekdays for scheduler

Can define multiple weekdays for scheduler to be allowed to run on.
Made parsing Day's more robust.

wsp-mbuckley committed:
Scheduler log message for found AVL database

wsp-mbuckley committed:
AVL database SQL chunksize parameter

Added environment variable for defining SQL chunksize to avoid memory errors when writing GTFS stop times data to AVL database.
Memory errors will shutdown the program to avoid continuously overloading a VM.

wsp-mbuckley committed:
Database schema and SQLite math extension

Fixed output PostgreSQL table schemas and names.

wsp-mbuckley committed:
FIxed run metadata database JSON parameters

Fixed saving the run parameters to the database as JSON instead of a string.

wsp-mbuckley committed:
Fixed filling delay calculation in DB

Added the delay calculation value when inserting timetable data into database.

hulsiejames commented on pyproject.toml:
Not seen this before but just had a quick google, looks like it's main job is the firsttime setup of the tool and correct interdependancies?

is this automated or done manually?

wsp-mbuckley replied:
`pyproject.toml` is a generic config file for Python projects, it includes settings for various development tools such as pylint, black, mypy and pytest and it also contains the package build settings.

The package build settings contain all the information required by pip to build and install a Python package, it also contains build metadata such as authors, descriptions and dependencies. The file is created and updated manually but shouldn't need updating that regularly, if people are cloning the git repo and setting up the repo for development they can still use the requirements files for installing the dependencies.

The pyproject.toml is used in the other CAF packages to allow them to be published on PyPI, BODSE hasn't been published on PyPI yet but can be installed with the git url using `pip install git+{git-url}@{branch-name}` so to install this branch would be `pip install git+https://github.com/Transport-for-the-North/BODS-Extractor.git@gtfs_adjustment`

hulsiejames commented on requirements.txt:
```diff
@@ -2,6 +2,13 @@ python>=3.11.0, <4.0.0
+# Required for converting AVL from lat / lon to easting / northing
+pyproj>=3.6.0, <4.0.0
+
+# Packages not available on conda-forge, only available on PyPI
```
Are arguments stating which package manager should be used provided by commandline,
or will this try cona-forge for everything and then just not install `gtfs-realtime-bindings`?

wsp-mbuckley replied:
'requirements.txt' files don't give any information around what package manager to use (conda-forge vs pip), there are other file formats which do. The 'requirements.txt' file just provides package names and versions (and in some cases a URL instead of version).

In this case if you attempt to install all packages with conda it will fail due to not finding the gtfs package, I have been commenting out that line, installing using conda then installing the final package using pip. I think it might be clearer if we split the requirements file into 2:
- 'requirements.txt': containing all packages available from pip or conda
- 'requirements-pip.txt': containing the packages which are only available from pip

Then creating an environment with conda could be done with the following 2 commands:
- `conda create -n env_name --file requirements.txt` to create the env and install the packages available on conda
- `pip install -r "requirements-pip.txt"` to install the remaining pip only packages

This could be done with pip (or pipenv) in a single command: `pip install -r requirements.txt -r requirement-pip.txt`

wsp-mbuckley replied:
Split into 2 requirement files in commit f39e0bb6da74409e35fb6e89593941fb07d76fd0

hulsiejames commented on src/bodse/avl/adjust.py:
```diff
@@ -0,0 +1,526 @@
+    ----------
+    stops : pd.DataFrame
+        DataFrame containing stop locations, requires columns
+        'stop_lat' and 'stop_lon' which contain the latitude
```
Could add v quick check that `stop_lat` & `stop_lon` are indeed wihtin `stops`, raising a missing column error if not?

wsp-mbuckley replied:
Done based on your suggestion in commit f39e0bb6da74409e35fb6e89593941fb07d76fd0

hulsiejames commented on src/bodse/avl/adjust.py:
```diff
@@ -0,0 +1,526 @@
+        appended, containing the easting and northing values. Any
+        translation errors will cause the new columns to contain NaNs.
+    """
+    eastnorth = pd.DataFrame(
```
```suggestion
    required_cols = ["stops_lat", "stops_lon"]
    missing_cols = [col for col in required_cols if col not in stops.columns]
    if missing_cols:
        raise ValueError(f"`stops` DataFrame is missing the required columns: {missing_cols}")
    
    eastnorth = pd.DataFrame(
```

wsp-mbuckley replied:
Done based on your suggestion in f39e0bb6da74409e35fb6e89593941fb07d76fd0

wsp-mbuckley committed:
New requirements file for PyPI only

Moved dependencies only available on PyPI into their own requirements file.
Added column check to adjustment stop coordinates translation.

wsp-mbuckley commented:
@hulsiejames addressed your suggestions, if your happy with everything I'll merge it?

hulsiejames commented on src/bodse/avl/adjust.py:
```diff
@@ -111,9 +111,14 @@ def _translate_stop_coordinates(stops: pd.DataFrame) -> pd.DataFrame:
-        stops[["stop_lat", "stop_lon"]]
-        .apply(lambda x: gtfs.lat_lon_to_bng(x["stop_lat"], x["stop_lon"]), axis=1)
+        stops[columns]
+        .apply(lambda x: gtfs.lat_lon_to_bng(x[columns[0]], x[columns[0]]), axis=1)
```
should this be
```
gtfs.lat_lon_to_bng(x[columns[**0**]], x[columns[**1**]]),
```
rather than
```
gtfs.lat_lon_to_bng(x[columns[**0**]], x[columns[**0**]]),
```

wsp-mbuckley replied:
Thanks for noting that, fixed in 344c064ad7bdb742882007689564c1cc5252eb69

wsp-mbuckley committed:
Fixed GTFS adjust coordinate translation bug

Also, commented python version in requirements file because it isn't necessary for conda and will break pip install.

hulsiejames commented on src/bodse/avl/raw.py:
```diff
@@ -0,0 +1,322 @@
+    journey_code: Optional[str] = None
+
+    # XML tag lookups
+    _expected_tags = {
```
should "latitiude" and "longitude" not be included here as `_expected_tags`?

Noting that they are not optional, and if they are missing they'd be ignored below by 
```
        for name, tag in cls._warning_tags.items():
            data[name] = _findtext(vehicle_journey, tag, parent, if_missing="ignore")
``` 

May be misunderstanding though!

wsp-mbuckley replied:
This file isn't actually used anymore, it was part of some tests looking at using the TransXChange format instead of GTFS. It was partially working for parsing the TXC format but we decided against using it while it was still WIP, might be worth me making that clearer in the module docstring that it's WIP and no currently used.

hulsiejames commented on src/bodse/scheduler.py:
```diff
@@ -0,0 +1,704 @@
+class SchedulerConfig(config_base.BaseConfig):
+    """Parameters for running BODSE scheduler."""
+
+    output_folder: types.DirectoryPath
```
A question mainly here.

is `types.DirectoryPath` interchangable with `pydantic.DirectoryPath`, noting that it would be `pydantic.tpyes.DirectoryPath` opposed to `pydantic.DirectoryPath`? 

Assming to be different from `pathlib.Path`, which can be a dir or file path, this seems to explicitly be a dir only path

wsp-mbuckley replied:
`pydantic.DirectoryPath` is an alias for `types.DirectoryPath` so they are referring to the exact same class. This is a wrapper class around `pathlib.Path` which checks that the path is to an existing directory and returns a `pathlib.Path` object. The type is purely used for additional validation

hulsiejames replied:
makes sense, cheers!

hulsiejames commented on src/bodse/scheduler.py:
```diff
@@ -0,0 +1,704 @@
+    teams_webhook_url: Optional[pydantic.HttpUrl] = None
+
+    run_months: int = pydantic.Field(2, ge=1, le=12)
+    run_month_day: int = pydantic.Field(5, ge=1, le=28)
```
Another question!

Assuming max day of 28 here is due to Feb. 

If there was to be support for running on later date in different month I assume this would require a new validator to check that the `run_month_day` would be a valid value given the month used? 

I don't think this is an issue though.

wsp-mbuckley replied:
Yeah so that it could run in Feb, if the days was above the number in the current month then it wouldn't be able to run within that month and therefore might need to wait an additional month

hulsiejames commented on src/bodse/teams.py:
```diff
@@ -0,0 +1,223 @@
+            Title of the post.
+        message : str
+            Main body of the post, should use markdown formatting.
+        color : str,  default "ffffff" (white)
```
Is this frame colour or text colour?

Only thinking if this is white text for teams in dark mode, would it show on teams not in dark mode

hulsiejames commented on src/bodse/timetable/download.py:
```diff
@@ -0,0 +1,215 @@
+        LOG.warning("cannot get feed version for non-existent file '%s'", path)
+        return None
+
+    if not zipfile.is_zipfile(path):
```
does `download_gtfs_feed` not also unzip archvies, so could try look for the unzipped archive if the zipped does not exist?

(Though, there shouldn't be a reason you'd have the unziped archived and not the zipped)

wsp-mbuckley replied:
The scheduler process which calls `get_feed_version` doesn't call `download_gtfs_feed` it calls `download_to_file` which just saves the GTFS zipfile to a folder and doesn't extract it. `download_to_file` isn't used as part of the scheduler process.

hulsiejames replied:
Ahh cheers, that makes sense. 

Apologies, I was looking down at main()

I'll stop for the day to avoid any more confusion 😄, think I've read too many lines for one afternoon!

wsp-mbuckley committed:
Upgraded to CTK 0.5 and Pydantic v2

Updated caf.toolkit to 0.5 and pydantic to v2.5.
Fixed pylint issues with pydantic.types module.
Updated pydantic classes to v2 validators.

wsp-mbuckley committed:
Fixed GTFS adjustment time columns

Fixed bug with decimal point in arrival and depature time columns of adjusted output e.g. 13.0:00 instead of 13:00.

MattBuckley-TfN committed:
Added seconds to adjusted GTFS outputs

Included seconds in timing columns in stop times output from GTFS adjusted.
Previously times output as HH:MM now output as HH:MM:SS.
